### DISCUSSION TITLE
Set environment variable based on git branch

### DISCUSSION THREAD

**Discussion by ralph** [community/community#25301] [Actions]
<p>Hi community,</p>
<p>I’m sorry if this has been asked before, but I did not find anything useful.</p>
<p>My situation is like this: I have a <code>develop</code> branch that corresponds to the staging env and a <code>master</code> branch that corresponds to the production env. Each time I push to one of the branches, GH actions runs to build/test/deploy. This works fine.</p>
<p>My problem is, that the workflow files for develop and master are very duplicated, see:</p>
<pre><code>➜ diff .github/workflows/{develop.yaml,master.yaml}
1c1
&lt; name: Build and deploy staging
---
&gt; name: Build and deploy production
5c5
&lt; - develop
---
&gt; - master
14c14
&lt; ENVIRONMENT: staging
---
&gt; ENVIRONMENT: production
</code></pre>
<p>The only things that differ are:</p>
<ul>
<li>the workflow name</li>
<li>the branch name (develop/master)</li>
<li>the value of the ENVIRONMENT variable (staging/production)</li>
</ul>
<p>I’d really like to consolidate this and have 1 file only for both staging &amp; production. So is there a way to run the workflow on pushes to both develop &amp; master branches? And to set an env var based on the branch that is pushed to?</p>
<p>Are there other ways to consolidate this? Many thanks for any suggestions. :slight_smile:</p>

**Comment by ssboisen** [comment]
<p>Have you considered just rendering the yaml files using a template and a small script? then whenever you make changes to the workflow you do it to the template and rerender</p>

**Comment by bomb-on** [comment]
<p>Did you try to use</p>
<pre><code>${GITHUB_REF##*/}
</code></pre>
<p>This should extract the branch part from “GITHUB_REF” env variable which is usually in form of “refs/heads/branch-name”.</p>

**Comment by ralph** [comment]
<p>Thank you <a class="mention" href="/bomb-on">@bomb-on</a> and <a class="mention" href="/ssboisen">@ssboisen</a>, I just took another look at the docs and found a syntax to set ENV vars that persist between steps:</p>
<p><a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-an-environment-variable-set-env" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-an-environment-variable-set-env</a></p>
<p>This is what I did to set the ENVIRONMENT variable automatically:</p>
<pre><code>steps:
    - name: Set env to staging
      if: endsWith(github.ref, '/develop')
      run: |
        echo "::set-env name=ENVIRONMENT::staging"
    - name: Set env to production
      if: endsWith(github.ref, '/master')
      run: |
        echo "::set-env name=ENVIRONMENT::production"
</code></pre>
<p>It’s kind of a simple mapping between branches and environment names and works in this case as we have only 2. When you run it, it looks like this:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/8/8ce5f7eadfb132e67255ee733d51a4ca548b0be8.png" data-download-href="/uploads/short-url/k6ry9Q2m5mVy6vJm46kcDvAWqda.png?dl=1" title="Bildschirmfoto 2019-12-03 um 20.16.19.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/2190/181090149-3566722e-9010-48e1-ac39-be481c99b61c.png" alt="Bildschirmfoto 2019-12-03 um 20.16.19.png" data-base62-sha1="k6ry9Q2m5mVy6vJm46kcDvAWqda" width="690" height="284" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/8/8ce5f7eadfb132e67255ee733d51a4ca548b0be8_2_690x284.png, https://aws1.discourse-cdn.com/github/original/2X/8/8ce5f7eadfb132e67255ee733d51a4ca548b0be8.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/8/8ce5f7eadfb132e67255ee733d51a4ca548b0be8.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/8/8ce5f7eadfb132e67255ee733d51a4ca548b0be8_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Bildschirmfoto 2019-12-03 um 20.16.19.png</span><span class="informations">882×364 20.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>See that the production step is skipped, as I was pushing to the “develop” branch. With this var set, everything else falls into place.</p>
<p>At the top of my workflow file I have the following now:</p>
<pre><code>name: Build and deploy
on:
  push:
    branches:
    - develop
    - master
</code></pre>
<p>I am happy now as I have to maintain 1 workflow file only and don’t have to remember to run a script when I change something. Thank you again for supporting me with your ideas.</p>

**Comment by ralph** [comment]
<p>Hi <a class="mention" href="/mklappir">@mklappir</a>, I’m doing the same thing pretty much already, as I’m reading from github.ref, which is pretty much the same as GITHUB_REF. Extracting the part after the last slash is a neat trick, but the endsWith macro serves me just as well.</p>
<p>Extracting the current branch name was just part of the problem, but my main concern was how to map the branch name (develop, master) to the environment name (staging, production) and save that part in an environment variable that persists between steps. So I need:</p>
<ul>
<li>“develop” branch :arrow_right: export ENVIRONMENT=“staging”</li>
<li>“master” branch :arrow_right: export ENVIRONMENT=“production”</li>
</ul>
<p>I agree that in an ideal world, I would just rename the environments according to the branch name, but that would require a lot of work on other ends which I want to avoid, especially now that I have a solution that works for me.</p>

**Comment by Notalifeform** [comment, marked as answer]
<p>Solution above works, but is being deprecated:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/" target="_blank" rel="noopener nofollow ugc">The GitHub Blog</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2019/03/blog-card.png?fit=5001%2C2626" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/" target="_blank" rel="noopener nofollow ugc">GitHub Actions: Deprecating set-env and add-path commands - GitHub Changelog</a></h3>

<p>GitHub Actions: Deprecating set-env and add-path commands</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>so it looks like this</p>
<pre><code class="lang-auto">steps:
    - name: Set env to staging
      if: endsWith(github.ref, '/develop')
      run: |
        echo "ENVIRONMENT=staging" &gt;&gt; $GITHUB_ENV
    - name: Set env to production
      if: endsWith(github.ref, '/master')
      run: |
       echo "ENVIRONMENT=production" &gt;&gt; $GITHUB_ENV
</code></pre>

**Comment by Simran-B** [comment]
<p>If you use PowerShell then this should do the trick:</p>
<pre><code class="lang-auto">run: |
  echo "ENVIRONMENT=staging" | Out-File -FilePath $Env:GITHUB_ENV -Encoding utf8 -Append
</code></pre>
<p>In Windows <code>cmd</code> it should work like this:</p>
<pre><code class="lang-auto">run: |
  chcp 65001 &gt; nul
  echo ENVIRONMENT=staging &gt;&gt; %GITHUB_ENV%</code></pre>

**Comment by haydenk** [comment]
<p><code>gitgub.ref_name</code> and <code>$GITHUB_REF_NAME</code> has the information you need without having to use <code>endsWith()</code></p>
<p>This is how I did it to get around updating the environment being deprecated.</p>
<pre><code class="lang-auto">jobs:
  setup-stage:
    outputs:
      env-name: ${{ steps.env-name.outputs.env }}
    runs-on: ubuntu-latest
    steps:
      - name: Setup Environment Name
        id: env-name
        run: echo "::set-output name=env::${{ fromJSON('{"main":"prod","develop":"dev"}')[github.ref_name] }}"
  build:
    needs: setup-stage
    runs-on: ubuntu-latest
    steps:
      - run: echo "Environment: ${{ needs.setup-stage.outputs.env-name }}"
</code></pre>

**Comment by Simran-B** [comment]
<p>FWIW, ref_name appears to be a recent addition. It was certainly not available at the time when endsWith was suggested.</p>
